feat(manifest): embed per-build metadata and publish unified firmware manifest

mrpollo opened the pull request:
This PR adds comprehensive firmware manifest metadata to PX4 builds, enabling better firmware discovery and management for tools like QGroundControl. Per-build metadata is generated from the defconfig and CMake board variables, embedded into each `.px4` file, and aggregated into a single unified manifest on S3 that QGC and other tools can consume.

## Changes

### 1. Board-level manifest generation
**Files:** `Tools/manifest/gen_board_manifest_from_defconfig.py`, `platforms/nuttx/CMakeLists.txt`

- Generates per-build manifest JSON with hardware and variant metadata
- Extracts architecture, USB vendor/product IDs, chip type, manufacturer from defconfig
- `variant` field sourced from CMake's `${LABEL}` (the `.px4board` filename stem, e.g. `multicopter.px4board` → `"multicopter"`)
- Integrated into CMake build system via `add_custom_command`

### 2. Manifest bundled into *.px4 files
**Files:** `Tools/px_mkfw.py`

- Each `.px4` file now contains a `manifest` object with board metadata
- Adds `manifest_version` and `sha256sum` fields
- `_merge_manifest` helper merges the per-build JSON fragment into the firmware descriptor
- Backward compatible: tested with QGC 4.4.5 and daily builds

### 3. Label and category fields
**Files:** `Kconfig`, `Tools/manifest/gen_board_manifest_from_defconfig.py`, `boards/px4/fmu-v6x/*.px4board`

- `CONFIG_BOARD_LABEL_PRETTY`: human-readable variant name shown in QGC (e.g. "Multicopter")
- `CONFIG_BOARD_FIRMWARE_CATEGORY`: override for auto-detected category (`vehicle`, `peripheral`, `dev`, `bootloader`)
- Auto-detection from variant label and `CONFIG_BOARD_ROMFSROOT` (`cannode` → peripheral)
- Phase 1 rollout on all `px4/fmu-v6x` variants; follow-up PR will cover the remaining ~258 `.px4board` files
- `artifact_type` discriminator (flat field, value `"px4"`) to allow future non-`.px4` producers (VOXL2 `.deb`, Linux tarballs) to coexist in the same manifest without a `format_version` bump

### 4. Unified firmware manifest
**Files:** `Tools/manifest/update_firmware_manifest.py`

- Scans all `*.px4` files in an artifacts directory and extracts their embedded metadata
- Fetches the existing unified manifest from S3 and upserts the current release inline
- Single file at `s3://px4-travis/Firmware/manifest.json` contains all releases and all builds for O(1) lookup
- Schema uses `format_version: 2` with `releases` as a dict keyed by version
- Auto-detects release channels (stable, beta, dev) from version string and tracks `latest_*` per channel

### 5. CI integration
**Files:** `.github/workflows/build_all_targets.yml`

On version tag releases:
1. Backs up the existing `manifest.json` from S3 as a workflow artifact (90-day retention)
2. Runs `update_firmware_manifest.py` to upsert this release into the unified manifest
3. Uploads the updated `manifest.json` to S3 and attaches it to the GitHub release

## Architecture

```
s3://px4-travis/Firmware/
├── manifest.json         # Unified: all releases × all builds, O(1) lookup
├── v1.15.0/
│   ├── px4_fmu-v6x_default.px4
│   ├── px4_fmu-v6x_multicopter.px4
│   └── ...
├── v1.16.1/
│   └── ...
└── stable/               # Symlink to latest stable
```

## Example Manifest

```json
{
  "format_version": 2,
  "updated_at": 1738890000,
  "latest_stable": "v1.16.1",
  "latest_beta":   "v1.17.0-beta1",
  "latest_dev":    "v1.17.0-alpha1",
  "releases": {
    "v1.16.1": {
      "channel": "stable",
      "release_date": "2026-01-15",
      "build_count": 170,
      "builds": [
        {
          "filename": "px4_fmu-v6x_multicopter.px4",
          "url": "https://github.com/PX4/PX4-Autopilot/releases/download/v1.16.1/px4_fmu-v6x_multicopter.px4",
          "sha256sum": "abc123...",
          "image_size": 1234567,
          "board_id": 53,
          "artifact_type": "px4",
          "manifest": {
            "name": "px4_fmu-v6x",
            "target": "px4_fmu-v6x_multicopter",
            "variant": "multicopter",
            "label_pretty": "Multicopter",
            "firmware_category": "vehicle",
            "manufacturer": "Holybro",
            "hardware": {
              "architecture": "arm",
              "chip": "stm32h753ii",
              "vendor_id": "0x3185",
              "product_id": "0x0038",
              "productstr": "Pixhawk 6X"
            }
          }
        }
      ]
    }
  }
}
```

A live test manifest with 5 real releases and 891 firmware builds is published at [px4-travis.s3.amazonaws.com/Firmware/manifest.json](https://px4-travis.s3.amazonaws.com/Firmware/manifest.json).

## Scope

Manifest generation currently runs only for NuttX targets (via `platforms/nuttx/CMakeLists.txt`). VOXL2 (qurt/posix), Linux SBCs, and SITL targets are not yet indexed; extending coverage to non-NuttX producers is a follow-up. The `artifact_type` discriminator is in place so those can land additively without a schema break.

## Related

- QGC companion PR: [mavlink/qgroundcontrol#13966](https://github.com/mavlink/qgroundcontrol/pull/13966)
- Follow-up: populate `CONFIG_BOARD_LABEL_PRETTY` across remaining ~258 `.px4board` files
- Follow-up: extend manifest producers to VOXL2 / Linux / SITL targets

dagar commented on Tools/ci/generate_manifest_json.py:
```diff
@@ -0,0 +1,273 @@
+        'architecture': '',
+    }
+    if boardtarget_file.endswith("default.px4board") or \
+        boardtarget_file.endswith("performance-test.px4board") or \
```
why?

dagar commented:
The easiest thing would be to generate all of this at build time where everything is easily available, but I think it would depend how and where we want to update the manifest.

If the manifest is updated by each build individually that would simplify a lot of things in PX4 itself, but then we'd have more to worry about in atomically updating the manifest.json in S3. Alternatively if we want to do bulk manifest.json updates in a single job  maybe we should push everything we possibly need into a .px4 file.

mrpollo commented:
I think this PR still needs to clarify under what circumstances the manifest is built and updated, and ensure those are documented somewhere, otherwise, we might be good for now.

Regarding the QGC implementation, here's the early draft https://github.com/mavlink/qgroundcontrol/pull/13966

mrpollo commented:
**New fields in manifest `label_pretty` and `firmware_category`**

Today, when QGC presents firmware options to a user, it shows raw build target names like `px4_fmu-v6x_default`, `px4_fmu-v6x_multicopter`, `px4_fmu-v6x_bootloader`. For experienced developers that's fine, but for regular users picking firmware for their drone it's cryptic — they shouldn't need to parse internal naming conventions to flash the right thing.

<img width="410" height="358" alt="Screenshot 2026-02-10 at 11 17 55 AM" src="https://github.com/user-attachments/assets/8a8b5b43-a6fc-4ac0-b8ca-d3041c2fd320" />

As we move toward per-vehicle-type firmware variants (multicopter, fixedwing, rover, etc.), the list of targets per board is growing. Without metadata, QGC has no way to know which builds are meant for end users, which are developer tools, which are sensor peripherals, and which are bootloaders — it has to show everything or resort to fragile filename string-matching.

These two new manifest fields solve both problems.

### `label_pretty`
A human-readable name for the build variant (e.g. "Multicopter", "Rover", "Default"). This is what QGC displays instead of raw target strings.

Set via `CONFIG_BOARD_LABEL_PRETTY` in each `.px4board` file. Due to Kconfig inheritance, **every variant must set its own value** — otherwise it inherits from `default.px4board`, which would be wrong.

### `firmware_category`
Classifies each build into one of four buckets:

| Value | What it means | QGC behavior |
|-------|--------------|-------------|
| `vehicle` | Production firmware for a vehicle type (multicopter, fixedwing, rover, etc.) | Shown to users by default |
| `peripheral` | Firmware for CAN sensor nodes (GPS, optical flow, magnetometer, distance sensor, etc.) | Shown in a dedicated peripheral/sensor section |
| `dev` | Developer/engineering builds (default, zenoh, mavlink-dev, performance-test, etc.) | Hidden by default, advanced mode only |
| `bootloader` | Bootloader binaries | Never shown to end users |

This is **auto-detected** — no manual config needed:
- Vehicle types are detected from the build label (`multicopter`, `fixedwing`, `vtol`, `rover`, `uuv`, `spacecraft`)
- Peripherals are detected from `CONFIG_BOARD_ROMFSROOT="cannode"` — all ~18 CAN sensor boards across ARK, Holybro, CUAV, Freefly, Matek, and NXP use this
- Bootloaders are detected from `bootloader`/`canbootloader` labels
- Everything else falls to `dev`

A **build-time warning** fires if an unrecognized label (not `default`) falls through to `dev`, so new vehicle types won't be silently hidden. Boards can always override via `CONFIG_BOARD_FIRMWARE_CATEGORY` in the `.px4board` file.

### What this means for hardware manufacturers

If you maintain a board in the PX4 tree:
- Add `CONFIG_BOARD_LABEL_PRETTY` to your `.px4board` files so QGC shows a clean name
- `firmware_category` auto-detects correctly — you don't need to touch it unless you have a non-standard variant name
- CAN peripheral boards (anything with `ROMFSROOT=cannode`) are automatically classified as `peripheral`

### Backward compatibility

Old `.px4` files without these fields still work — QGC falls back to raw label / filename matching (existing behavior). These are purely additive fields, no `format_version` bump needed.

### Scope

Right now only `px4/fmu-v6x` (10 variants) has `CONFIG_BOARD_LABEL_PRETTY` set. There are ~258 `.px4board` files across 111 boards, with 17 boards that have vehicle-type variants and ~18 CAN peripheral boards.

**Question for reviewers:** Should we add `CONFIG_BOARD_LABEL_PRETTY` to all ~258 board files in this PR, or land the infrastructure now and do a bulk update in a follow-up? It's scriptable — the label is just a title-cased version of the filename in most cases — but it's a large diff that touches every board and might benefit from a separate review. Leaning toward a follow-up PR so this one stays focused on the manifest plumbing.

dakejahl commented:
What is this waiting on? Can we get this and the QGC PR across the finish line soon?

julianoes commented:
I just stumbled on this as well. Yes, this or something like that would be great!

mrpollo committed:
feat(manifest): embed per-build metadata and publish unified firmware manifest

Generate per-build manifest metadata from defconfig and CMake board
variables, embed it into each .px4 file, and aggregate everything into a
single unified firmware manifest on S3 for ground stations like QGC.

- Board-level manifest generation (gen_board_manifest_from_defconfig.py):
  architecture, USB VID/PID, chip, manufacturer, and variant (from the
  .px4board filename stem) extracted per build.
- px_mkfw.py embeds a manifest object plus manifest_version and a
  sha256sum of the (signed) firmware image into every .px4 descriptor.
- CONFIG_BOARD_LABEL_PRETTY and CONFIG_BOARD_FIRMWARE_CATEGORY Kconfig
  options; firmware_category auto-detected from the variant label and
  ROMFSROOT ("cannode" -> peripheral), overridable per board.
- update_firmware_manifest.py scans built .px4 files and upserts the
  release into a single unified manifest.json (format_version 2, releases
  keyed by version) for O(1) lookup across all releases.
- CI publishes/updates s3://px4-travis/Firmware/manifest.json on tagged
  releases, backing up the previous manifest first.
- artifact_type discriminator for future non-.px4 producers.

Phase 1 sets CONFIG_BOARD_LABEL_PRETTY on the px4/fmu-v6x variants only.

Signed-off-by: Ramon Roche <mrpollo@gmail.com>

Claudio-Chies commented on Tools/manifest/gen_board_manifest_from_defconfig.py:
```diff
@@ -0,0 +1,144 @@
+from typing import Dict
+
+_VEHICLE_LABELS = frozenset({
+    "multicopter", "fixedwing", "vtol", "rover", "uuv", "spacecraft",
```
where would builds for "niche" vehicles go? e.g baloon or autogyro. would it make sense to also have an "other" label?

dakejahl replied:
Agreed, we need an other/misc category. The "vehicle" designation also might be too narrow, these are **build targets** not necessarily vehicles. I need to take a closer look after https://github.com/PX4/PX4-Autopilot/pull/27667

sfuhrer commented on docs/en/hardware/firmware_manifest.md:
```diff
@@ -0,0 +1,110 @@
+
+The firmware category is automatically inferred from the build label (the last segment of the target string after the final `_`):
+
+- Labels `multicopter`, `fixedwing`, `vtol`, `rover`, `uuv`, `spacecraft` → `vehicle`
```
```suggestion
- Labels `copter`, `fixedwing`, `vtol`, `rover`, `uuv`, `spacecraft` → `vehicle`
```
To include helicopter.

julianoes replied:
Wouldn't helicopter be its own thing?

hamishwillee replied:
I don't know the answer to this, but if it goes in, we need to modify `multicopter` elsewhere in this doc.

sfuhrer reviewed (commented):
This PR doesn't remove the "default" builds yet, correct? Only once we do that it will actually make a difference, so I guess we'll then do a follow-up once this new workflows have been validated?

dakejahl commented:
> This PR doesn't remove the "default" builds yet, correct? Only once we do that it will actually make a difference, so I guess we'll then do a follow-up once this new workflows have been validated?

I'm working on it here, if you want to take a preliminary review
https://github.com/PX4/PX4-Autopilot/pull/27667

dakejahl committed:
feat(manifest): align vehicle labels with the target-class taxonomy

Rename "multicopter" to "copter" and add "airship" in _VEHICLE_LABELS to
match the target-class build redesign's variant names; "fixedwing" is kept over
"plane". Without "airship", airship boards mis-detect as "dev".

dakejahl committed:
feat(manifest): derive firmware_category from the resolved target class

The target-class build redesign (PX4/PX4-Autopilot#27667) makes
CONFIG_TARGET_CLASS_<X> the authoritative description of a build's role, so the
manifest should read it instead of reverse-engineering the role from the variant
label. Pass PX4_TARGET_CLASS into the generator and map it to firmware_category
(vehicle classes + linux -> vehicle; cannode + io -> peripheral; sitl + ros2 ->
dev); keep the variant-label and cannode-ROMFS heuristics as a fallback for
boards not yet on the target-class grammar, and let CONFIG_BOARD_FIRMWARE_CATEGORY
still override.

This is forward-compatible: PX4_TARGET_CLASS is empty until the redesign lands,
so behavior on this branch is unchanged. Once merged it fixes the io coprocessor
(was misclassified 'dev', now 'peripheral') and the linux SBC flight controllers,
and removes the need for any voxl2 special-case.

Signed-off-by: Jacob Dahl <dahl.jakejacob@gmail.com>

dakejahl commented:
I pushed two commits to align this PR with https://github.com/PX4/PX4-Autopilot/pull/27667. Once https://github.com/PX4/PX4-Autopilot/pull/27667 is in we can re-name the .px4boards here.

julianoes commented on .github/workflows/build_all_targets.yml:
```diff
@@ -303,4 +307,57 @@ jobs:
+      - name: Backup Existing Firmware Manifest
+        if: startsWith(github.ref, 'refs/tags/v')
+        run: |
+          s3_base="https://px4-travis.s3.us-west-1.amazonaws.com/Firmware"
+          mkdir -p manifest_backup
+          curl -sf "${s3_base}/manifest.json" -o manifest_backup/manifest.json.backup || echo "No existing manifest to backup"
+          if [ -f manifest_backup/manifest.json.backup ]; then
+            echo "Backed up existing firmware manifest ($(wc -c < manifest_backup/manifest.json.backup) bytes)"
```
For tags, could we just add the tag to the filename, and then have a copy or link to the latest one? That way we keep old manifests without awkward backup files without version info.

julianoes commented on boards/px4/fmu-v6x/mavlink-dev.px4board:
```diff
@@ -1,2 +1,3 @@
+CONFIG_BOARD_LABEL_PRETTY="MAVLink Dev"
```
```suggestion
CONFIG_BOARD_LABEL_PRETTY="MAVLink development dialect"
```

julianoes commented on platforms/nuttx/CMakeLists.txt:
```diff
@@ -475,6 +474,33 @@ if (TARGET parameters_xml AND TARGET airframes_xml)
+		COMMAND
+			${PYTHON_EXECUTABLE} ${PX4_SOURCE_DIR}/Tools/manifest/gen_board_manifest_from_defconfig.py
+				--defconfig	${NUTTX_DEFCONFIG}
+				--manufacturer	"${CONFIG_CDCACM_VENDORSTR}"
```
```suggestion
				--vendorstr	"${CONFIG_CDCACM_VENDORSTR}"
```

To keep it in line with productstr and vid/pid.

julianoes commented on Tools/manifest/gen_board_manifest_from_defconfig.py:
```diff
@@ -0,0 +1,168 @@
+    "sitl": "dev", "ros2": "dev",
+}
+
+def parse_defconfig(path: str) -> Dict[str, str]:
```
This needs a comment. It's not clear what it is doing.

julianoes commented on Tools/manifest/gen_board_manifest_from_defconfig.py:
```diff
@@ -0,0 +1,168 @@
+            d[k.strip()] = v
+    return d
+
+def norm_hex(s: str) -> str:
```
Also not clear.

julianoes commented on Tools/manifest/gen_board_manifest_from_defconfig.py:
```diff
@@ -0,0 +1,168 @@
+    ap = argparse.ArgumentParser(description="Generate board manifest (prefer CMake-passed overrides, fallback to defconfig).")
+    ap.add_argument("--defconfig", required=False, help="Path to defconfig (fallback only)")
+    # explicit overrides coming from CMake
+    ap.add_argument("--manufacturer", default="")
```
```suggestion
    ap.add_argument("--vendorstr", default="")
```

julianoes commented on Tools/manifest/gen_board_manifest_from_defconfig.py:
```diff
@@ -0,0 +1,168 @@
+
+    defcfg = parse_defconfig(args.defconfig) if args.defconfig else {}
+
+    manufacturer    = pick(args.manufacturer, "CONFIG_BOARD_MANUFACTURER", defcfg)
```
What's this? I don't see `CONFIG_BOARD_MANUFACTURER` anywhere in the code.

Shouldn't this come from `CONFIG_CDCACM_VENDORSTR` which is actually already what is used as the cli argument, so this seems to be duplicate paths to the same thing.

julianoes commented on Tools/manifest/gen_board_manifest_from_defconfig.py:
```diff
@@ -0,0 +1,168 @@
+    defcfg = parse_defconfig(args.defconfig) if args.defconfig else {}
+
+    manufacturer    = pick(args.manufacturer, "CONFIG_BOARD_MANUFACTURER", defcfg)
+    productstr      = pick(args.productstr,   "CONFIG_BOARD_PRODUCTSTR",   defcfg)
```
Same here, that's already `CONFIG_CDCACM_PRODUCTSTR`.

Unless, this is some sort of fallback for boards without USB like cannodes?

julianoes commented on Tools/manifest/gen_board_manifest_from_defconfig.py:
```diff
@@ -0,0 +1,168 @@
+        firmware_cat = detect_firmware_category(variant, defcfg, target=target,
+                                                target_class=target_class)
+
+    manifest = {
+        "name": name,
+        "target": target,
+        "variant": variant,
+        "label_pretty": label_pretty,
+        "firmware_category": firmware_cat,
+        "manufacturer": manufacturer,
+        "hardware": {
+            "architecture": arch,
+            "vendor_id": vid,
+            "product_id": pid,
+            "chip": chip,
+            "productstr": productstr
+        }
+    }
```
```suggestion
    manifest = {
        "name": name,
        "target": target,
        "variant": variant,
        "label_pretty": label_pretty,
        "firmware_category": firmware_cat,
        "hardware": {
            "architecture": arch,
            "vendor_id": vid,
            "product_id": pid,
            "chip": chip,
             "vendorstr": vendorstr,
            "productstr": productstr
        }
    }
```

julianoes commented on Tools/px_mkfw.py:
```diff
@@ -63,8 +65,18 @@ def mkdesc():
 	proto['build_time']	= 0
 	proto['image']		= bytes()
 	proto['image_size']	= 0
+	proto['sha256sum'] = ""
```
What's this for? Just to check binary integrity?
Because we also have signing which would be different.

hamishwillee commented on docs/en/config/firmware.md:
```diff
@@ -78,6 +78,15 @@ If _QGroundControl_ installs the FMUv2 target (see console during installation),
+## Firmware Variants
+
+PX4 boards may provide multiple firmware variants for different vehicle types (multicopter, fixed-wing, rover, etc.).
+QGroundControl shows vehicle firmware by default.
```
```suggestion
QGroundControl shows default vehicle firmware by default.
```

hamishwillee commented on docs/en/hardware/board_support_guide.md:
```diff
@@ -113,6 +113,9 @@ Contact PX4 board maintainers at [boards@px4.io](mailto:boards@px4.io) and reque
 2. The assignment of REV and VER ID resistor values.
 3. If the board supports USB: Either request the assignment of a USB VID and PID or provide the USB VID and PID.
```
Perhaps a note that assignment of USB VID/PID is a member benefit.

hamishwillee commented on docs/en/hardware/board_support_guide.md:
```diff
@@ -113,6 +113,9 @@ Contact PX4 board maintainers at [boards@px4.io](mailto:boards@px4.io) and reque
 2. The assignment of REV and VER ID resistor values.
 3. If the board supports USB: Either request the assignment of a USB VID and PID or provide the USB VID and PID.
 
+4. Set `CONFIG_BOARD_LABEL_PRETTY` in all `.px4board` variant files so that ground stations can display a human-readable name for each build variant.
```
A "for example" might by nice.


```suggestion
4. Set `CONFIG_BOARD_LABEL_PRETTY` in all `.px4board` variant files so that ground stations can display a human-readable name for each build variant.
   For example, the default variant label is always `default`, and `zenoh` might be used for a variant that includes zenoh binaries.
```

hamishwillee replied:
Actually this contradicts the other doc " **Set `CONFIG_BOARD_FIRMWARE_CATEGORY="vehicle"`** in the `.px4board` file — use this for one-off overrides or when a board variant doesn't follow the standard naming."

So you only do this in some cases right?

hamishwillee committed:
docs(docs): prettier

hamishwillee commented on docs/en/hardware/firmware_manifest.md:
```diff
@@ -0,0 +1,110 @@
+  "target": "px4_fmu-v6x_multicopter",
+  "label_pretty": "Multicopter",
+  "firmware_category": "vehicle",
+  "manufacturer": "Holybro",
```
Would have thought this should be associaed with the hardware.

hamishwillee commented on docs/en/hardware/firmware_manifest.md:
```diff
@@ -0,0 +1,110 @@
+
+### Auto-Detection
+
+The firmware category is automatically inferred from the build label (the last segment of the target string after the final `_`):
```
Consistent naming is important. Before it is a variant label, now i ti s abuild label.


```suggestion
The firmware category is automatically inferred from the variant label (the last segment of the target string after the final `_`):
```

hamishwillee commented on docs/en/hardware/firmware_manifest.md:
```diff
@@ -0,0 +1,110 @@
+
+The firmware category is automatically inferred from the build label (the last segment of the target string after the final `_`):
+
+- Labels `multicopter`, `fixedwing`, `vtol`, `rover`, `uuv`, `spacecraft` → `vehicle`
```
I would invert - the arrow confused me. So for all of these

```suggestion
- `vehicle`: Labels `multicopter`, `fixedwing`, `vtol`, `rover`, `uuv`, `spacecraft`
```

EDIT ~Would be good to have exhaustive lists, or links to how you can find out the options, or a default option if you don't know what to use.~ Ah, I see below.

hamishwillee commented on docs/en/hardware/firmware_manifest.md:
````diff
@@ -0,0 +1,110 @@
+The `label_pretty` field provides a human-readable name for each build variant.
+Ground stations display this instead of raw target strings like `px4_fmu-v6x_default`.
+
+### Setting `label_pretty`
+
+Set `CONFIG_BOARD_LABEL_PRETTY` in each `.px4board` file:
+
+```text
+CONFIG_BOARD_LABEL_PRETTY="Multicopter"
+```
````
Why? You've just said I can set it in the manifest, so why do it here? I suspect that it is to override the field, but I don't know why you'd need to since it isn't a category that is used for controlling display like the firmware category (I assume).

EDIT Ah, I see you answer why it needs to be set in the next section, but this _still doesn't explain why you need to set it_ in the manifest and in the board.
